transport.go

```diff
@@ -202,6 +224,49 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
 	return conn, nil
 }
 
+func (t *transport) holePunch(ctx context.Context, network string, addr *net.UDPAddr) (tpt.CapableConn, error) {
+	pconn, err := t.connManager.Dial(network, addr)
+	if err != nil {
+		return nil, err
+	}
+	defer pconn.DecreaseCount()
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	connCh := make(chan tpt.CapableConn)
+
+	key := addr.String()
+	t.holePunchingMx.Lock()
+	t.holePunching[key] = activeHolePunch{connCh: connCh, ctx: ctx}
```

Comment:
useless context?

Reply:
Also, what if there are _two_ hole-punch attempts? We should probably error out in that case.

Reply:
Removed the context.
Why should two hole-punch attempts be an error?

Reply:
(discussed out of band)